### ISSUE TITLE
Panic while fetching drm fdinfo

### ISSUE THREAD

**Issue by orhun** [ulissesf/qmassa#3]
Hey!

Great project, I wanted to give it a try but hit this error:

```rust
thread 'main' panicked at src/drm_fdinfo.rs:240:23:
index out of bounds: the len is 1 but the index is 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'd want to contribute a fix but I'm not sure what `drm_fdinfo` is and how to circumvent the error :)

**Comment by ulissesf** [comment]
Hi @orhun, thanks for the report and for testing! Sorry about the issue. I'm guessing you don't have an Intel GPU, do you? I haven't been able to test it with an AMD one yet, but it shouldn't break (I thought). Do you mind sending me the backtrace? I'll take a look later tonight to fix it and probably put in README the current vendor support more clearly.

**Comment by kaydenil** [comment]
Hey @ulissesf, I tried this on my AMD machine and was able to reproduce it.  I hacked in a debug! call to print the lines being split:

[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-vram:\t16700 KiB"
[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-gtt: \t8192 KiB"   
[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-cpu: \t0 KiB"

In other words, a couple of the values have a pesky extra space after the colon, so the splitting on ":\t" isn't working.

```diff
-            let kv: Vec<&str> = l.split(":\t").collect();
+            let kv: Vec<&str> = l.split(":").collect();
             let k = kv[0];
-            let v = kv[1];
+            let v = kv[1].trim();
```

seems to get it past this, though it might split too much if there are colons on the right hand side.

Once up, however, it doesn't display any process, memory, or frequency info, but presumably that's just amdgpu support missing.

**Comment by kaydenl** [comment]
Hey @ulissesf, I tried this on my AMD machine and was able to reproduce it.  I hacked in a debug! call to print the lines being split:

[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-vram:\t16700 KiB"
[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-gtt: \t8192 KiB"   
[2024-10-31T18:01:02Z DEBUG qmassa::drm_fdinfo] LINE: "drm-memory-cpu: \t0 KiB"

In other words, a couple of the values have a pesky extra space after the colon, so the splitting on ":\t" isn't working.

```diff
-            let kv: Vec<&str> = l.split(":\t").collect();
+            let kv: Vec<&str> = l.split(":").collect();
             let k = kv[0];
-            let v = kv[1];
+            let v = kv[1].trim();
```

seems to get it past this, though it might split too much if there are colons on the right hand side.

Once up, however, it doesn't display any process, memory, or frequency info, but presumably that's just amdgpu support missing.

**Comment by ulissesf** [comment]
Hi @kaydenl, thanks for debugging it! I got one AMD card borrowed to look at this over the weekend. I'll get your changes and some others in this area so we don't create more strings than needed and trim key/value to be more robust. 

Yes, amdgpu support is missing, but I'll add the basics this weekend.

**Comment by orhun** [comment]
Hey guys, yup, I have an AMD GPU, sorry for not making clear while submitting the issue. Glad to hear that you're going to add support for it!

**Comment by ulissesf** [comment]
@orhun and @kaydenl, I've merged a fix and a couple of other improvements to get as much data from amdgpu's DRM fdinfo files right now. Let me know if it works for you as well (I tested with an AMD RX 7600M tonight). I will add some of the amdgpu driver-specific info over the weekend.

I went with a different fix, @kaydenl, as we indeed can have more colons in some of the key-value info (e.g. drm-pdev). Let me know what you think.
